express-error-handling/server/index.js

```
@@ -0,0 +1,146 @@
+  const sql = `
+    insert into "grades" ("name", "course", "score")
+    values ($1, $2, $3)
+    returning *
+  `;
+  const params = [name, course, score];
+  db.query(sql, params)
+    .then(result => {
+      const [newGrade] = result.rows;
+      res.status(201).json(newGrade);
+    })
+    .catch(err => next(err));
+});
+
+app.get('/api/grades/:gradeId', (req, res, next) => {
+  const gradeId = Number(req.params.gradeId);
+  if (!Number.isInteger(gradeId) || gradeId < 1) {
+    res.status(400).json({
+      error: 'grade must be a positive integer'
+    });
+    return;
+  }
+  const sql = `
+    select *
+      from "grades"
+     where "gradeId" = $1
+  `;
+  const params = [gradeId];
+  db.query(sql, params)
+    .then(result => {
+      const [grade] = result.rows;
+      if (!grade) {
+        throw new ClientError(404, `cannot find grade with gradeId ${gradeId}`);
+      } else {
+        res.json(grade);
+      }
+    })
+    .catch(err => next(err));
+});
+
+app.put('/api/grades/:gradeId', (req, res, next) => {
+  const gradeId = Number(req.params.gradeId);
+  if (!Number.isInteger(gradeId) || gradeId < 1) {
+    throw new ClientError(400, 'grade must be a positive integer');
+  }
+  const { name, course } = req.body;
+  const score = Number(req.body.score);
+  if (!Number.isInteger(score) || score < 0 || score > 100) {
+    throw new ClientError(400, 'score must be an integer between 0 and 100');
+  }
+  if (!name || !course) {
+    throw new ClientError(400, 'name, course, and score are required fields');
+  }
+  const sql = `
+    update "grades"
+       set "name"   = $1,
+           "course" = $2,
+           "score"  = $3
+     where "gradeId" = $4
+    returning *
+  `;
+  const params = [name, course, score, gradeId];
+  db.query(sql, params)
+    .then(result => {
+      const [updatedGrade] = result.rows;
+      if (!updatedGrade) {
+        res.status(404).json({
+          error: `cannot find grade with gradeId ${gradeId}`
+        });
```

Should throw a client error here